src/parser/jobs/dnc/modules/FeatherGauge.tsx

```diff
@@ -35,55 +38,64 @@ export default class FeatherGauge extends Module {
 
 	private feathersConsumed = 0
 	private avgGenerated = 0
-	private history: any[] = []
+	private history: Array<{t: number, y: number, isGenerator: boolean}> = [{t: 0, y: 0, isGenerator: false}]
 	private currentFeathers = 0
 	private featherOvercap = 0
 
 	protected init() {
-		this.addHook('cast', {by: 'player', abilityId: FEATHER_GENERATORS}, this.onCastGenerator)
+		this.addHook('aoedamage', {by: 'player', abilityId: FEATHER_GENERATORS}, this.onCastGenerator)
 		this.addHook('cast', {by: 'player', abilityId: FEATHER_CONSUMERS}, this.onConsumeFeather)
 		this.addHook('death', {to: 'player'}, this.onDeath)
 		this.addHook('complete', this.onComplete)
 	}
-	private onCastGenerator() {
+	private onCastGenerator(event: AoeEvent) {
+		if (!event.successfulHit) {
+			return
+		}
 		this.avgGenerated += FEATHER_GENERATION_CHANCE
-		this.setFeather(this.currentFeathers + FEATHER_GENERATION_CHANCE)
+		this.setFeather(this.currentFeathers + FEATHER_GENERATION_CHANCE, true)
 	}
 	private onConsumeFeather() {
 		this.feathersConsumed++
+
 		// If we consumed a feather when we think we don't have one, clearly we do, so update the history to reflect that
-		// TODO: count how many feathers have been spent since the last possible generation event and endure the history for
-		// those n events is correct.
 		if (this.currentFeathers < 1) {
-			const prevHistory = this.history.pop()
-			prevHistory.y = 1
-			this.history.push(prevHistory)
+			// Add the underrun amount to all events back to the previous spender so the graph shows we had enough to spend
+			const lastGeneratorIndex = _.findLastIndex(this.history, event => event.isGenerator)
```

Comment:
[Array.lastIndexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/lastIndexOf) 😉 

no biggie though, lodash is great

Reply:
couldn't figure out a simple way to use lastIndexOf to find the index where a given property of the objects in the array is a value (true in this case). All of the solutions I found were needlessly complicated, so I went with lodash instead since it has what I wanted out of the box.